### ISSUE TITLE
Enable PyCharm debugger to automatically attach to worker processes by patching startup command using PyDev

### ISSUE THREAD

**Issue by nirvana-msu** [ray-project/ray#14005]
### Describe your feature request

Currently, [Local mode](https://docs.ray.io/en/master/starting-ray.html#local-mode) is the only feasible way to debug Ray worker/actor code, but it would be great if debugging worked  in non-local mode out of the box when using PyCharm.

PyCharm patches (see e.g. description [here](https://youtrack.jetbrains.com/issue/PY-32703#focus=Comments-27-3170212.0-0)) all Python functions that are used to start new processes when launched in Debug mode (using [pydev_monkey.monkey_patch_module](https://github.com/JetBrains/intellij-community/blob/8e73fa69fde14293feef0f37296e4567e21797c4/python/helpers/pydev/_pydev_bundle/pydev_monkey.py#L393-L398)) - so whenever you launch new subprocesses in Python (including using e.g. `multiprocessing`), debugger attaches automatically - and everything works seamlessly.

The reason it doesn't work with Ray is because worker processes/actors are started by [Cython / native code](https://github.com/ray-project/ray/blob/982c606b86e9d2b7009e3e38c40d8b313148b568/python/ray/_raylet.pyx#L1146), bypassing PyDev patches. If you were to patch the Python startup command within Cython code whenever PyDev is used (same as how PyCharm does it for respective Python functions) then debugging should just magically start working - and you wouldn't even need Local mode anymore, if you use PyCharm.

**Comment by pcmoritz** [comment]
Hey @nirvana-msu have you had a look at https://docs.ray.io/en/master/ray-debugging.html?

Let us know if that solution works for you!

**Comment by nirvana-msu** [comment]
@pcmoritz yeah I've seen that, but I don't want to be "dropped into a PDB session". What I want is an actual interactive debugging in PyCharm where I can step through the code line-by-line and use all the debugging abilities that come with IDE. This is infinitely more useful and more user-friendly than pdb. And it generally works perfectly out of the box, including for sub-processes launched from Python - and should work with Ray workers as well if we were to patch python worker launch command accordingly (to invoke PyDev).

**Comment by gduvauchelle** [comment]
@nirvana-msu This is not as magic as the solution you propose but I have managed to debug code executed remotely with ray (without local mode) by opening a reverse ssh tunnel to the machine where the code is executed, and using the [remote debugging](https://www.jetbrains.com/help/pycharm/remote-debugging-with-product.html#remote-debug-config) feature of pycharm.

**Comment by austinnichols101** [comment]
This seemingly small change would make a *huge difference* in my workflow.  

The PyCharm debugger is outstanding, but unfortunately useless with Ray actor code.  I can debug everything except for the actors which is a huge pain.

**Comment by Suke0811** [comment]
I tried manually attaching a Pycharm debugger using the following code. It appears to get connected but the debugger still skips breakpoints inside a ray actor. I'm fine manually attaching a debugger, but I want them to at least work. 

```python
i=ray.init()
ip, port = i.get('address').split(':')
pydevd_pycharm.settrace(ip, port=22000, stdoutToServer=True,
                        `stderrToServer=True)
```
>>> 
venv/bin/python /snap/pycharm-professional/280/plugins/python/helpers/pydev/pydevd.py --multiprocess --qt-support=auto --client 127.0.0.1 --port 40555 --file /test/actor_debugger.py
Connected to pydev debugger (build 221.5080.212)

**Comment by mgerstgrasser** [comment]
> @nirvana-msu This is not as magic as the solution you propose but I have managed to debug code executed remotely with ray (without local mode) by opening a reverse ssh tunnel to the machine where the code is executed, and using the [remote debugging](https://www.jetbrains.com/help/pycharm/remote-debugging-with-product.html#remote-debug-config) feature of pycharm.

@gduvauchelle Could I ask you how exactly you set this up? In particular, how you figured out which port to connect the debugger to? I see over two dozen ports opened by Ray, but none of the ones I've tried to connect to seem to be a python debugger.

**Comment by Vozf** [comment]
I've managed to debug ray actor with pycharm. Here are my steps. I use port 6067
1. set up reverse ssh port forwarding `ssh -R localhost:6067:localhost:6067 <name_of_machine>`
2. [Start](https://www.jetbrains.com/help/pycharm/remote-debugging-with-product.html#remote-debug-config) debug server on your local machine. Use hostname `localhost` and port 6067. You will also need to set up project path mapping so you see the code, but you can do it later or use Auto-download when the breakpoint is triggered
3. Follow the steps described on debug server creation:
On remote env: `pip install pydevd-pycharm~=241.14494.241`
Modify your code(I've ran it in the constructor of ray actor that needs to be debugged)
```
import pydevd_pycharm
pydevd_pycharm.settrace('localhost', port=6067, stdoutToServer=True, stderrToServer=True)
```
4. Start debug server in pycharm
5. Run the script you want to debug

**Comment by liangjh2001** [comment]
> I've managed to debug ray actor with pycharm on remote machine. I guess you can also skip port forwarding to do it on local machine. Here are my steps. I use port 6067
> 
> 1. set up reverse ssh port forwarding `ssh -R localhost:6067:localhost:6067 <name_of_machine>`
> 2. [set up](https://www.jetbrains.com/help/pycharm/remote-debugging-with-product.html#remote-debug-config) debug server on your local machine. Use hostname `localhost` and port 6067. You will also need to set up project path mapping so you see the code, but you can do it later or use Auto-download when the breakpoint is triggered
> 3. Follow the steps described on debug server creation:
>    On remote env: `pip install pydevd-pycharm~=241.14494.241`(Version should be different for you, the exact version is shown in debug server configuration)
>    Modify your code(I've ran it in the constructor of ray actor that needs to be debugged)
> 
> ```
> import pydevd_pycharm
> pydevd_pycharm.settrace('localhost', port=6067, stdoutToServer=True, stderrToServer=True)
> ```
> 
> 4. Start debug server in pycharm
> 5. Run the script you want to debug. Breakpoints should now trigger

I successfully debugged the Ray framework code (verl) on a remote server using PyCharm, although it was a bit complex. Thank you!
